packages/babel-plugin-transform-classes/src/transformClass.js

```diff
@@ -371,21 +338,66 @@ export default function transformClass(
   }
 
   function verifyConstructor() {
-    if (!classState.isDerived) return;
+    const { userConstructorPath: path, isDerived, file } = classState;
+    if (!path) return;
 
-    const path = classState.userConstructorPath;
     const body = path.get("body");
 
-    path.traverse(findThisesVisitor);
-
-    let guaranteedSuperBeforeFinish = !!classState.bareSupers.size;
-
+    let thisAlias;
     let thisRef = function() {
-      const ref = path.scope.generateDeclaredUidIdentifier("this");
+      const ref = path.scope.generateUidIdentifier("this");
       thisRef = () => t.cloneNode(ref);
+      path.scope.push({
+        id: ref,
+        init: (thisAlias = !isDerived ? t.thisExpression() : null),
+      });
       return ref;
     };
 
+    // prevent `assertThisInitialized(this)` duplicates
+    let assertThisInited = false;
+
+    // Fix issue #8335, traverse all of the `this` expressions with assertion and aliases
+    path.traverse(
+      verifyConstructorVisitor(path, {
```

Comment:
This recreates the visitor every time a class is transformed. Can you hoist it up so that it is created only once?

Reply:
y, already hoist it up in the latest commit.